### PR TITLE
Fix tmpdir on non root become

### LINKED ISSUE #41340: 2.6.0rc1: blockinfile with become_user has permissions failures in temporary directory
##### SUMMARY
When blockinfile, with become_user, is used to create a new file, the module fails due to permissions failures in the temporary directory

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

blockinfile

##### ANSIBLE VERSION

```
ansible 2.6.0rc1
  config file = None
  configured module search path = [u'/home/centos/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

No changes (ansible-config --dump returns nothing)

##### OS / ENVIRONMENT

CentOS 7.4

##### STEPS TO REPRODUCE

inventory.ini
```yaml
target ansible_connection=local
```

playbook.yml
```yaml
---
- hosts: all
  tasks:
  - blockinfile:
      create: yes
      block: |
        export foo=bar
      marker: "# {mark} Here there be a marker"
      dest: /home/testuser/testing.txt
      owner: testuser
      group: testuser
      mode: 0644
    become_user: testuser
    become: yes
````

Ensure /home/testuser/testing.txt does not exist.

Run ansible-playbook -i inventory.ini playbook.yml -vvv

##### EXPECTED RESULTS

I expected success, like I get with 2.5.x:

```
ansible-playbook 2.5.3
  config file = None
  configured module search path = [u'/home/centos/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
No config file found; using defaults
Parsed /home/centos/inventory.ini inventory source with ini plugin

PLAYBOOK: playbook.yml *************************************************************************************************************************************
1 plays in playbook.yml

PLAY [all] *************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************
task path: /home/centos/playbook.yml:2
Using module file /usr/lib/python2.7/site-packages/ansible/modules/system/setup.py
<target> ESTABLISH LOCAL CONNECTION FOR USER: centos
<target> EXEC /bin/sh -c 'echo ~centos && sleep 0'
<target> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195 `" && echo ansible-tmp-1528489378.63-85098428369195="` echo /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195 `" ) && sleep 0'
<target> PUT /home/centos/.ansible/tmp/ansible-local-27028e8Ph1H/tmplZgegG TO /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195/setup.py
<target> EXEC /bin/sh -c 'chmod u+x /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195/ /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195/setup.py && sleep 0'
<target> EXEC /bin/sh -c '/usr/bin/python /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195/setup.py && sleep 0'
<target> EXEC /bin/sh -c 'rm -f -r /home/centos/.ansible/tmp/ansible-tmp-1528489378.63-85098428369195/ > /dev/null 2>&1 && sleep 0'
ok: [target]
META: ran handlers

TASK [blockinfile] *****************************************************************************************************************************************
task path: /home/centos/playbook.yml:4
Using module file /usr/lib/python2.7/site-packages/ansible/modules/files/blockinfile.py
<target> ESTABLISH LOCAL CONNECTION FOR USER: centos
<target> EXEC /bin/sh -c 'echo ~centos && sleep 0'
<target> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/tmp/ansible-tmp-1528489379.28-171541100059794 `" && echo ansible-tmp-1528489379.28-171541100059794="` echo /var/tmp/ansible-tmp-1528489379.28-171541100059794 `" ) && sleep 0'
<target> PUT /home/centos/.ansible/tmp/ansible-local-27028e8Ph1H/tmpKW0ISr TO /var/tmp/ansible-tmp-1528489379.28-171541100059794/blockinfile.py
<target> EXEC /bin/sh -c 'setfacl -m u:testuser:r-x /var/tmp/ansible-tmp-1528489379.28-171541100059794/ /var/tmp/ansible-tmp-1528489379.28-171541100059794/blockinfile.py && sleep 0'
<target> EXEC /bin/sh -c 'sudo -H -S -n -u testuser /bin/sh -c '"'"'echo BECOME-SUCCESS-qskrbdreonavsqmsainpewhljtntukqq; /usr/bin/python /var/tmp/ansible-tmp-1528489379.28-171541100059794/blockinfile.py'"'"' && sleep 0'
<target> EXEC /bin/sh -c 'rm -f -r /var/tmp/ansible-tmp-1528489379.28-171541100059794/ > /dev/null 2>&1 && sleep 0'
changed: [target] => {
    "changed": true,
    "diff": [
        {
            "after": "",
            "after_header": "/home/testuser/testing.txt (content)",
            "before": "",
            "before_header": "/home/testuser/testing.txt (content)"
        },
        {
            "after_header": "/home/testuser/testing.txt (file attributes)",
            "before_header": "/home/testuser/testing.txt (file attributes)"
        }
    ],
    "invocation": {
        "module_args": {
            "attributes": null,
            "backup": false,
            "block": "export foo=bar\n",
            "content": null,
            "create": true,
            "delimiter": null,
            "dest": "/home/testuser/testing.txt",
            "directory_mode": null,
            "follow": false,
            "force": null,
            "group": "testuser",
            "insertafter": null,
            "insertbefore": null,
            "marker": "# {mark} Here there be a marker",
            "marker_begin": "BEGIN",
            "marker_end": "END",
            "mode": 420,
            "owner": "testuser",
            "path": "/home/testuser/testing.txt",
            "regexp": null,
            "remote_src": null,
            "selevel": null,
            "serole": null,
            "setype": null,
            "seuser": null,
            "src": null,
            "state": "present",
            "unsafe_writes": null,
            "validate": null
        }
    },
    "msg": "File created"
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************
target                     : ok=2    changed=1    unreachable=0    failed=0
```

##### ACTUAL RESULTS

```
ansible-playbook 2.6.0rc1
  config file = None
  configured module search path = [u'/home/centos/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
No config file found; using defaults
Parsed /home/centos/inventory.ini inventory source with ini plugin

PLAYBOOK: playbook.yml *************************************************************************************************************************************
1 plays in playbook.yml

PLAY [all] *************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************
task path: /home/centos/playbook.yml:2
<target> ESTABLISH LOCAL CONNECTION FOR USER: centos
<target> EXEC /bin/sh -c 'echo ~centos && sleep 0'
<target> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207 `" && echo ansible-tmp-1528489241.67-154074287834207="` echo /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/system/setup.py
<target> PUT /home/centos/.ansible/tmp/ansible-local-26937OUx1qh/tmpsaSW45 TO /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207/setup.py
<target> EXEC /bin/sh -c 'chmod u+x /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207/ /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207/setup.py && sleep 0'
<target> EXEC /bin/sh -c '/usr/bin/python /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207/setup.py && sleep 0'
<target> EXEC /bin/sh -c 'rm -f -r /home/centos/.ansible/tmp/ansible-tmp-1528489241.67-154074287834207/ > /dev/null 2>&1 && sleep 0'
ok: [target]
META: ran handlers

TASK [blockinfile] *****************************************************************************************************************************************
task path: /home/centos/playbook.yml:4
<target> ESTABLISH LOCAL CONNECTION FOR USER: centos
<target> EXEC /bin/sh -c 'echo ~centos && sleep 0'
<target> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/tmp/ansible-tmp-1528489242.66-190869334332659 `" && echo ansible-tmp-1528489242.66-190869334332659="` echo /var/tmp/ansible-tmp-1528489242.66-190869334332659 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/files/blockinfile.py
<target> PUT /home/centos/.ansible/tmp/ansible-local-26937OUx1qh/tmp4y0GYj TO /var/tmp/ansible-tmp-1528489242.66-190869334332659/blockinfile.py
<target> EXEC /bin/sh -c 'setfacl -m u:testuser:r-x /var/tmp/ansible-tmp-1528489242.66-190869334332659/ /var/tmp/ansible-tmp-1528489242.66-190869334332659/blockinfile.py && sleep 0'
<target> EXEC /bin/sh -c 'sudo -H -S -n -u testuser /bin/sh -c '"'"'echo BECOME-SUCCESS-dkqkzvhlmkxabaonclqypkdsoabveqpw; /usr/bin/python /var/tmp/ansible-tmp-1528489242.66-190869334332659/blockinfile.py'"'"' && sleep 0'
<target> EXEC /bin/sh -c 'rm -f -r /var/tmp/ansible-tmp-1528489242.66-190869334332659/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_c67IEZ/ansible_module_blockinfile.py", line 351, in <module>
    main()
  File "/tmp/ansible_c67IEZ/ansible_module_blockinfile.py", line 335, in main
    write_changes(module, result, real_path)
  File "/tmp/ansible_c67IEZ/ansible_module_blockinfile.py", line 163, in write_changes
    tmpfd, tmpfile = tempfile.mkstemp(dir=getattr(module, 'tmpdir', None))
  File "/usr/lib64/python2.7/tempfile.py", line 304, in mkstemp
    return _mkstemp_inner(dir, prefix, suffix, flags)
  File "/usr/lib64/python2.7/tempfile.py", line 239, in _mkstemp_inner
    fd = _os.open(file, flags, 0600)
OSError: [Errno 13] Permission denied: '/var/tmp/ansible-tmp-1528489242.66-190869334332659/tmpXemXXr'

fatal: [target]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_c67IEZ/ansible_module_blockinfile.py\", line 351, in <module>\n    main()\n  File \"/tmp/ansible_c67IEZ/ansible_module_blockinfile.py\", line 335, in main\n    write_changes(module, result, real_path)\n  File \"/tmp/ansible_c67IEZ/ansible_module_blockinfile.py\", line 163, in write_changes\n    tmpfd, tmpfile = tempfile.mkstemp(dir=getattr(module, 'tmpdir', None))\n  File \"/usr/lib64/python2.7/tempfile.py\", line 304, in mkstemp\n    return _mkstemp_inner(dir, prefix, suffix, flags)\n  File \"/usr/lib64/python2.7/tempfile.py\", line 239, in _mkstemp_inner\n    fd = _os.open(file, flags, 0600)\nOSError: [Errno 13] Permission denied: '/var/tmp/ansible-tmp-1528489242.66-190869334332659/tmpXemXXr'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
        to retry, use: --limit @/home/centos/playbook.retry

PLAY RECAP *************************************************************************************************************************************************
target                     : ok=1    changed=0    unreachable=0    failed=1
```

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY
also avoid exception if tmpdir and remote_tmp are None

fixes #41340, #38058

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

remote temp
##### ANSIBLE VERSION

```
>=2.6
```

### CODE REVIEW

File: lib/ansible/module_utils/basic.py [hunk 1]
```diff
@@ -930,9 +930,10 @@ def __init__(self, argument_spec, bypass_checks=False, no_log=False,
 
     @property
     def tmpdir(self):
-        # if _ansible_tmpdir was not set, the module needs to create it and
-        # clean it up once finished.
-        if self._tmpdir is None:
+        # if _ansible_tmpdir was not set and we have a remote_tmp,
+        # the module needs to create it and clean it up once finished.
+        # otherwise it stays None and tempfile consumer will use system dir defaults
+        if self._tmpdir is None and self._remote_tmp:
```
Comment (abadger) [inline lib/ansible/module_utils/basic.py:936]: I think we need _remote_tmp here.  So we should either make sure we're always passing in a _remote_tmp or we should set a default here.  (Since we're already setting a default in action/__init__.py:_update_module_args, it feels like we should set this there as well).

File: lib/ansible/plugins/action/__init__.py [hunk 2]
```diff
@@ -665,7 +665,9 @@ def _update_module_args(self, module_name, module_args, task_vars):
         module_args['_ansible_keep_remote_files'] = C.DEFAULT_KEEP_REMOTE_FILES
 
         # make sure all commands use the designated temporary directory if created
-        module_args['_ansible_tmpdir'] = self._connection._shell.tmpdir
+        # FIXME: deal with non root become, temporarily set to None and let remote_tmp take care of it
+        # module_args['_ansible_tmpdir'] = self._connection._shell.tmpdir
+        module_args['_ansible_tmpdir'] = None
```
Comment (abadger) [inline lib/ansible/plugins/action/__init__.py]: Heh, I just meant this code to show that setting _ansible_tmpdir = None was what we needed to do.  I think that it's a bit too big a change to set this to None unconditionally.  Probably better to revert the changes to how we call tempfile.* until we figure out how to implement this only for privilege descalation.

File: lib/ansible/module_utils/basic.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/module_utils/basic.py:936]: its stays None, which when consumed by tempfile will use system dirs, using default might not work as `~/.ansible/tmp` can break things and user changed it for a reason

File: lib/ansible/plugins/action/__init__.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/plugins/action/__init__.py]: just as easy to remove it, i left it like this to force fixing, which is what im looking into now, posted this JIC we want to go forward for RC, this is not final fix

File: lib/ansible/module_utils/basic.py (hunk 1 above)
Comment (abadger) [reply lib/ansible/module_utils/basic.py:936]: Hmmm... okay.   None is an implementation detail, though.  So what we probably want is if we aren't given a remote_tmp either, then we should create a tmpdir in the system tempdir, store that in self._tmpdir and return it from this function.

Comment (bcoca) [reply lib/ansible/module_utils/basic.py:936]: so basically do the work tempfile already does?

File: test/integration/targets/remote_temp/aliases [hunk 3]
```diff
@@ -0,0 +1 @@
+posix/ci/group3
```
Comment (abadger) [inline test/integration/targets/remote_temp/aliases:1]: the target dir should be remote_tmp

File: test/integration/targets/remote_temp/playbook.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    - block:
+        - blockinfile:
+              create: yes
+              block: |
+                export foo=bar
+              marker: "# {mark} Here there be a marker"
+              dest: /tmp/testing.txt
+              mode: 0644
+        - file: path=/tmp/testing.txt state=absent
+      become_user: nobody
```
Comment (abadger) [inline test/integration/targets/remote_temp/playbook.yml]: nobody often has a home dir that can't be written to.  Perhaps the playbook has to create a user for this test?

File: lib/ansible/module_utils/basic.py (hunk 1 above)
Comment (abadger) [reply lib/ansible/module_utils/basic.py:936]: Yes.  That's the expectation if self._tmpdir is set to None when this property is accessed.

File: lib/ansible/plugins/action/__init__.py [hunk 5]
```diff
@@ -86,6 +89,15 @@ def run(self, tmp=None, task_vars=None):
                                  ' the tmpdir']
         del tmp
 
+        try:
+            self._admin_users = self._connection._shell.get_option('admin_users')
+            remote_user = self._connection.get_option('remote_user')
+        except AnsibleError:  # old custom plugins w/o get_option
+            self._admin_users = ['root']
+            remote_user = self._play_context.remote_user
```
Comment (abadger) [inline lib/ansible/plugins/action/__init__.py]: It seems like we need to do each of these separately?  So that if a plugin doesn't have that option we only fallback for that option, not both options?  (And that case will throw KeyError, not AnsibleError iirc).

Comment (abadger) [reply lib/ansible/plugins/action/__init__.py]: It also doesn't seem right to set these as self values in run().  It seems like we should either set these in __init__() if the action plugin is never reused or pass the values to the functions that use them.

File: lib/ansible/module_utils/basic.py (hunk 1 above)
Comment (abadger) [reply lib/ansible/module_utils/basic.py:936]: This is still needed.

File: lib/ansible/modules/files/assemble.py [hunk 6]
```diff
@@ -204,7 +204,7 @@ def main():
     if validate and "%s" not in validate:
         module.fail_json(msg="validate must contain %%s: %s" % validate)
 
-    path = assemble_from_fragments(src, delimiter, compiled_regexp, ignore_hidden)
+    path = assemble_from_fragments(src, delimiter, compiled_regexp, ignore_hidden, getattr(module, 'tmpdir', None))
```
Comment (abadger) [inline lib/ansible/modules/files/assemble.py:207]: Why are we using getattr everywhere instead of just module.tmpdir?  getattr being non-standard makes people paranoid about making changes but AFAICS, there is no reason to make people afraid of touching this code.

File: lib/ansible/plugins/action/__init__.py [hunk 7]
```diff
@@ -86,6 +89,17 @@ def run(self, tmp=None, task_vars=None):
...
+        # set info needed to make decisions, fallback for old custom plugins w/o get_option
+        try:
+            self._admin_users = self._connection._shell.get_option('admin_users')
+        except AnsibleError:
+            self._admin_users = ['root']
+        try:
+            remote_user = self._connection.get_option('remote_user')
+        except AnsibleError:
+            remote_user = self._play_context.remote_user
+        self._deescalate = bool(self._play_context.become and self._play_context.become_user not in self._admin_users + [remote_user])
```
Comment (abadger) [inline lib/ansible/plugins/action/__init__.py:101]: Setting these things here still feels wrong.  can you walk me through where something isn't being "finalized" when __init__ is run so that I can figure out why this is needed and if there's some other way that might be better?

File: lib/ansible/modules/files/assemble.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/modules/files/assemble.py:207]: mostly to allow people to use the module on older ansible core versions, I tend to err on the side of keeping it compatible if i can

File: lib/ansible/plugins/action/__init__.py (hunk 7 above)
Comment (bcoca) [reply lib/ansible/plugins/action/__init__.py:101]: the better way involves some rewrites in task_executor, which i'm planning, but they are out of scope for this PR

File: lib/ansible/modules/files/assemble.py (hunk 6 above)
Comment (abadger) [reply lib/ansible/modules/files/assemble.py:207]: K.  I think we should use module.tmpdir then.

**Comment by abadger** [comment]
An updated version of this was merged in https://github.com/ansible/ansible/pull/42396
